url47e6aca666_pocorall_automaton_tgz-pJ8-net_pocorall_automaton_DatatypesJ8/src/net/pocorall/automaton/BasicOperations.java

```diff
@@ -0,0 +1,622 @@
+	 */
+	static public DefaultAutomaton repeat(DefaultAutomaton a, int min) {
+		if (min == 0)
+			return repeat(a);
+		List<DefaultAutomaton> as = new ArrayList<DefaultAutomaton>();
+		while (min-- > 0)
+			as.add(a);
+		as.add(repeat(a));
+		return concatenate(as);
+	}
+
+	/**
+	 * Returns an automaton that accepts between <code>min</code> and
+	 * <code>max</code> (including both) concatenated repetitions of the
+	 * language of the given automaton.
+	 * <p/>
+	 * Complexity: linear in number of states and in <code>min</code> and
+	 * <code>max</code>.
+	 */
+	static public DefaultAutomaton repeat(DefaultAutomaton a, int min, int max) {
+		if (min > max)
+			return makeEmpty();
+		max -= min;
+		a.expandSingleton();
+		DefaultAutomaton b;
+		if (min == 0)
+			b = makeEmptyString();
+		else if (min == 1)
+			b = a.clone();
+		else {
+			List<DefaultAutomaton> as = new ArrayList<DefaultAutomaton>();
+			while (min-- > 0)
+				as.add(a);
+			b = concatenate(as);
+		}
+		if (max > 0) {
+			DefaultAutomaton d = a.clone();
+			while (--max > 0) {
+				DefaultAutomaton c = a.clone();
+				for (State p : c.getAcceptStates())
+					p.addEpsilon(d.initial);
+				d = c;
+			}
+			for (State p : b.getAcceptStates())
+				p.addEpsilon(d.initial);
+			b.deterministic = false;
+			b.clearHashCode();
+			b.checkMinimizeAlways();
+		}
+		return b;
+	}
+
+	/**
+	 * Returns a (deterministic) automaton that accepts the complement of the
+	 * language of the given automaton.
+	 * <p/>
+	 * Complexity: linear in number of states (if already deterministic).
+	 */
+	static public DefaultAutomaton complement(DefaultAutomaton a) {
+		a = a.cloneExpandedIfRequired();
+		BasicOperations.determinize(a);
+		a.totalize();
+		for (State p : a.getStates()) {
+			if (p.accept == null) {
+				p.accept = Object.class;
+			} else {
+				p.accept = null;
+			}
+		}
+		a.removeDeadTransitions();
+		return a;
+	}
+
+	/**
+	 * Returns a (deterministic) automaton that accepts the intersection of
+	 * the language of <code>a1</code> and the complement of the language of
+	 * <code>a2</code>. As a side-effect, the automata may be determinized, if not
+	 * already deterministic.
+	 * <p/>
+	 * Complexity: quadratic in number of states (if already deterministic).
+	 */
+	static public DefaultAutomaton minus(DefaultAutomaton a1, DefaultAutomaton a2) {
+		if (a1.isEmpty() || a1 == a2)
+			return makeEmpty();
+		if (a2.isEmpty())
+			return a1.cloneIfRequired();
+		if (a1.isSingleton()) {
+			if (a2.run(a1.singleton) != null)
+				return makeEmpty();
+			else
+				return a1.cloneIfRequired();
+		}
+		return intersection(a1, a2.complement());
+	}
+
+	/**
+	 * Returns an automaton that accepts the intersection of
+	 * the languages of the given automata.
+	 * Never modifies the input automata languages.
+	 * <p/>
+	 * Complexity: quadratic in number of states.
+	 */
+	static public DefaultAutomaton intersection(DefaultAutomaton a1, DefaultAutomaton a2) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 19. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.